Added conan pack to conan

sigmoidal opened the pull request:
This is an adapted version of the conan pack command described in #2025 for conan v0.29.1. There are some bug fixes also and it has been tested (lightly) on win and linux.

Would appreciate feedback on how it would make sense to improve.

ovidiub13 commented:
Why aren't you doing this PR against `develop`?

ovidiub13 commented on conans/client/command.py:
```diff
@@ -124,7 +124,57 @@ def new(self, *args):
+        try:
+            reference = ConanFileReference.loads(args.reference)
+        except ConanException:
+            print("Could not load reference: %s" % str(args.reference))
```
Use functions from the [Conan output system](https://github.com/conan-io/conan/blob/develop/conans/client/output.py) to print messages. Or use the `ConanException`

sigmoidal replied:
this was a conan-independent tool. I will switch it eventually, thx.

ovidiub13 commented on conans/client/command.py:
```diff
@@ -124,7 +124,57 @@ def new(self, *args):
+                                            version=version,
+                                            user=user,
+                                            channel=channel,
+                                            package_id=args.package_id,
```
You already have `name`, `version`, `user` and `channel` in `reference` which you pass. Why not extract them as needed in `pack()` and not pass so many arguments?

See `upload` command.

sigmoidal replied:
It is there because initially a reference was not used, rather a bunch of parameters.

ovidiub13 commented on conans/client/command.py:
```diff
@@ -124,7 +124,57 @@ def new(self, *args):
+                else:
+                    print("\n%s is not an existing package of %s" % (args.package_id, str(reference)))
+            else:
+                print("Could not load reference: %s" % str(args.reference))
```
Again, use Conan printing functions.

ovidiub13 commented on conans/client/conan_api.py:
```diff
@@ -429,7 +429,18 @@ def export_pkg(self, path, name, channel, source_folder=None, build_folder=None,
+                                 
+    @api_method
+    def pack(self, reference, name, version, user, channel,
+             package_id, properties, pack_name, output_dir=".", target_scheme=None):
```
Too many arguments. Same issue as above.

sigmoidal replied:
Thanks a lot for the feedback. 

I'd appreciate if we do the review top-down, i.e. starting from the user experience, rather than the code implementation. 

Is the functionality exposed to the user appropriate?
Does it cover the most common use cases?
Should we add any other parameters or rename some?

I will fix the code as you propose in the next iteration, thanks for looking into it.

memsharded replied:
I agree with this, lets focus on what it does first, code implementation second. Of course, style and other comments are good (specially the one about using the conan ``output``), but it would be very nice if you tried the functionality, did it work, the overall approach is good or not, just add a brief comment to the review: "everything looks good in this feature, just suggested a few implementation code changes". Many thanks to both! :)

memsharded commented:
I have changed the PR to develop. PR are not done to releases after they are released, in either case, a minor branch (0.29.2) should be done, and this is only for bugfixes. The normal process is PR to develop. Thanks!

memsharded commented:
I have reviewed this PR and my main concern is how this satisfies user expectations, especially regarding transitive dependencies. If you have an application or library A (In PkgA) that depends, lets say on shared libraries B (in PkgB), are users having to create manually "packs" for both? Then to use them, have to unzip the packs, manually move them to the same location to be found, or defining something in the system?

While this might work for 1 level and a very low number of dependencies (1-2), IMHO the process doesn't scale, and I think users would look at this feature for something to be able to distribute their final products, but this "packing" has to be complete and automated. I think this is quite easy to achieve with the ``imports`` or ``deploy`` functionalities, but then the process is different, because the final thing being packed is an amalgamation of several packages done in user space, not directly a package.

Wdyt?

ovidiub13 commented:
What I would need is a way for my final product to be archived so I can deliver it to the clients. I'd prefer this to be done automatically from the `conan create` command, and controlled via a flag that gives it the location (just an idea, not sure of it yet).

Using the `conan pack` command sounds good, but I find it hard to use in automatized environments, since the package ID needs to be captured with some scripts from the `conan create` command, otherwise I have no idea what it is. Also the reference.

I'd use this (after somehow getting the package ID and reference) after calling `conan create` on a Linux and a Windows machine, to package the artifacts, and then archive them together to be delivered to the client.

In a manual scenario, this is perfect.

Regarding scalability of usage ( @memsharded ) I don't see this command as a separate delivery methods for dependencies, but a a method of packing your final artifact in order to be delivered to the clients.

memsharded commented:
> Regarding scalability of usage ( @memsharded ) I don't see this command as a separate delivery methods for dependencies, but a a method of packing your final artifact in order to be delivered to the clients.

Yes, I agree, this is a method to pack your final artifact to be delivered to clients. This is why I am concerned about transitive dependencies, the current approach does not bundle the transitive dependencies, thus, I am concerned about its general applicability.

I am thinking of something that could in fact be more automated, something in the line:

```bash
$ conan install <some-pkg-ref> -pr=myprofile 
# <some-pkg> has a deploy() method, thus copy to local user folder all the artifacts, including 
# transitive dependencies binaries
$ conan pack . --name="myname" # so it works over the local copy
```
Typically, the output name could be very well the name of the profile used to install it. In this way, there is no explicit usage of any package ID at all, because it is computed from the profile at install time. I also think it is very explicit and easy for future "packers", for example, a Windows installer with InnoSetup might need an extra folder to do its stuff, which I feel it is better to have as a subfolder of the place where the artifacts where "deployed", as it is not clean to have those folders in the conan local cache.

sigmoidal commented:
Thank you both for all your comments and definitely for looking into it.

I recall someone told me to use a fixed release branch instead of the develop branch before I set off to do this, so I didn't know, but it makes sense, I will always use "develop" from now on.

I have to admit what you describe were the reasons I was so reluctant to turn my initial class into a conan command and I thought this is the job of an external tool. If you check, I did implement it using deploy initially, but then it started to get unnecessarily complicated. What if someone hasn't implemented deploy(), I still wanted to pack arbitrary packages that the author hasn't put any effort/thought about packing.

I think we have chances of finding a nice way to go about this. Let me try to add a couple of thoughts:

1) Dependencies: As a packager/end-user I would expect to be able to do BOTH actually. Be able to pack my own artifacts only, or my own artifacts + all deps. That could possibly be addressed using imports/ deploy .. and maybe a command line switch to define --with-deps <depth> (optionally, default infinite recursion?)

2) Package ID: When I implemented this through deploy(), I was initially not using much of the conan facilities, exactly to avoid the complexity of knowing what the package contains. Dumb script, but effective, because you could potentially script where you want it to pack from, etc, and it would do the job. I introduced package_id because it can be coupled with the output of conan search, this is how I identified what I want to package. The searching functionality with queries (-q) can be built into the conan pack command, so it is a one-step process, and instead of displaying the package id, it could feed them into the package_id option of "pack" as a list (so you can pack more things at once).

I think we have to insist to implement something reasonable like this and not give up on it. Simply because it is extremely useful and the use case is not going away, imo. I don't have strong feelings about implementing this as a conan command or not, but either way it would be helpful to brainstorm on what we want it to do, if it is feasible and finally how to approach it.

thank you.

memsharded commented:
I am ok with being it an external tool too, but I thought it is something that could be useful for quite a few other conan users, so it would be more convenient for them if it is integrated, and it will be easier to get feedback and usage for this tool.

I also agree that the command/tool could do both, it could just "pack" one package contents, or it could also do transitive dependencies. The reasons I think doing it via deploy() could be more convinient is:

- Usually packages contain many other things than the final binaries you want to re-distribute (headers, debug info, etc)
- Paths issues. ``deploy()`` will copy everything in the same place, with the desired folder layout, which makes much easier to automate usage (PATHs, etc)

sigmoidal commented:
I am not sure if it should be external or internal. Can you comment about the idea of using conan search functionality (-q) to identify and process the right package? why not?

The way you have in mind with deploy(), is it different from the way I initially did the Deployment() class, in which users were expected to call it with a one-liner in deploy() to get the job done?

If we go down this route of using deploy(), as a user I would like to do something simple:

1) If there is no deploy method specified, it should pack using "package/<hash>", just like it does now. With little effort.

2) if there is a deploy method, it should be extremely easy for the user to have the same functionality as in (1), but be able to supplement the "pack" with the additional files one may want, so there is need for some compositional behavior.

I am not sure how these can be implemented right now (limited conan api experience), some pointers would be useful:

a) recursive transient deps - any code chunks in the sources I should focus on?  
b) deploy -- how do I know it is missing from the user implementation. should deploy be called recursively on deps if we use it?

memsharded commented:
> Can you comment about the idea of using conan search functionality (-q) to identify and process the right package? why not?

I think it could be more useful than using the raw packageID, but I see some issues, as the packageID is not defined only by the settings/options (which might be used in the query), but also by the transitive dependencies. So for the same query you might get more than one binary, one for each different version of the dependencies they link with. It is more about the dependencies again. ``conan search`` cannot be used for that, for computing the full dependency graph, you need to input the complete configuration (profile, or settings, options, etc). 

So in the general case, I'd say that "packing" requires the full graph resolution, and thus, to be input full configuration, and transitive dependencies computed, to get the right packageID.

The way I think it is better to tackle this approach for scalability, future maintenance and extensibility is "divide and conquer". We have 2 different problems here:

1. From a given dependency graph, get all the artifacts we want. It is the end user who specify which artifacts to extract from the dependency graph, and there is no need that packages explicitly define anything regarding this.

2. Given a set of artifacts, binaries and other files, create a redistributable: zip, msi, .deb, or any other format

Point 2 is where most of the work is to be done, and the core of your PR is indeed implementing the creation of a compressed file from a set of artifacts. The other redistributables formats require more amount of work, because they are more complex. This point 2 can and should be implemented fully agnostic of any dependency graph, or package manager. I can already state the problem: "ey, I have here in this folder one exe, 2 dlls, a license and a data file, I need to create an MSI, a Deb package and a ZIP redistributable to put them in my webpage so users can download them"

Conan design always tries to decouple things as much as possible, to always let users do what they want. For example:

```cmake
include(... /conanbuildinfo.cmake)
conan_basic_setup()
```

The ``include()`` call could easily contain a call to do the setup automatically. But it is much better to do it in 2 steps, because users can more easily hack into it, and customize the setup at will. Same idea of doing ``git add . & git commit & git push`` instead of some ``git do-it`` command that does everything.

This idea might be equally valid for this case. If we try to do everything automagically, just one conan command, and voila, you get the final redistributable, that would probably block a very large number of users that needs their customization, maybe even outside conan. I can very well see users wanting to automate their redistributables like:

```bash
$ conan install <pkg-ref> -pr=profile # with a deploy() method
$ copy -R /some/path/images ./deployed/logos
$ conan pack . --output=MyPkgInstaller.msi
```

So, being purist, I'd say that point 1) functionality has to (and is already implemented by conan), and point2) functionality should be an external tool. But being pragmatic and as a convenience for users, I am very fine to integrate point 2) as a command in conan. But the code design has to be in such a way that it should be very easy to be a standalone application that knows nothing about conan at all. It should be something similar to what we did with the python linter for recipes. Indeed the linter is an utility that could/should be external: developers should run their python linters to check if their recipes are good. But it is very convenient if integrated and automated. But you can check the source code for ``export_linter.py``: almost no dependency to Conan at all, just an exception and a color type. Probably we could remove them too, and it will take just 3 minutes to extract and reuse that code outside of conan for another project.

Sorry for the long dissertation :) :) what do yo think?
